dev_guide/templates.adoc

```diff
@@ -239,9 +239,172 @@ application at will.
+      "iconClass": "icon-php"<4>
+    }
+  }
+  ...
+}
+----
+<1> The name of the template as it will appear to users.
+<2> A description of the template.
+<3> Tags to be associated with the template for searching and grouping.
+<4> An icon to be displayd with your template in the web console.
+====
+
+[[writing-labels]]
+=== Labels
+Templates can include a set of labels.  These labels will be added to each object created when the template is instantiated.  Defining a label in this way makes it easy for users to find and manage all the objects created from a particular template.
+
+====
+----
+{
+  "kind": "Template",
+  "apiVersion": "v1",
+  ...
+  "labels": {
+    "template": "cakephp-mysql-example"<1>
+  }
+  ...
+}
+----
+<1> A label that will be applied to all objects created from this template.
+====
+
+[[writing-parameters]]
+=== Parameters
+Parameters allow a value to be supplied by the user or generated when the template is instantiated.  This is useful for generating random passwords or allowing the user to supply a hostname or other user specific value that is required to customize the template.  Parameters can be referenced by placing the value "${PARAMETER_NAME}" in place of any string field in the template.
+
+====
+----
+{
+  "kind": "Template",
+  "apiVersion": "v1",
+  ...
+    {
+      "kind": "BuildConfig",
+      "apiVersion": "v1",
+      "metadata": {
+        "name": "cakephp-mysql-example",
+        "annotations": {
+          "description": "Defines how to build the application"
+        }
+      },
+      "spec": {
+        "source": {
+          "type": "Git",
+          "git": {
+            "uri": "${SOURCE_REPOSITORY_URL}",<1>
+            "ref": "${SOURCE_REPOSITORY_REF}"
+          },
+          "contextDir": "${CONTEXT_DIR}"
+        },
+      }
+    }
+   ...
+ "parameters": [
+    {
+      "name": "SOURCE_REPOSITORY_URL",<2>
+      "description": "The URL of the repository with your application source code",<3>
+      "value": "https://github.com/openshift/cakephp-ex.git"<4>
+      "required": true <5>
+    },
+	{
+      "name": "GITHUB_WEBHOOK_SECRET",
+      "description": "A secret string used to configure the GitHub webhook",
+      "generate": "expression",<6>
+      "from": "[a-zA-Z0-9]{40}"<7>
+    },
+  ]
+  ...
+}
+----
+<1> This value will be replaced with the value of the SOURCE_REPOSITORY_URL parameter when the template is instantiated.
+<2> The name of the parameter.  This value is displayed to users and used to reference the parameter within the template.
+<3> A description of the parameter.
+<4> A default value for the parameter which will be used if the user does not override the value when instantiating the template.
+<5> Indicates this parameter is required, meaning the user cannot override it with an empty value and if the parameter does not provide a default or generated value, the user must supply a value.
+<6> A parameter which has its value generated via a regular expression-like syntax.
```

Comment:
added, though really that content should move out of the architecture section and into this doc...  the architecture section should be reserved for high level overview of concept, the actual implementation/usage specifics should be in the admin and developer guides.

@bfallonf @adellape @CowboysFan fyi, something to consider as you're reworking the docs.